content/posts/using-the-redis-allocator-in-rust.md

```diff
@@ -0,0 +1,153 @@
+
+After a bit of searching, we find that this is the way Rust does name mangling: Unlike in C, and similarly to C++, in Rust multiple functions with the same name can coexist, since there are various namespace mechanisms such as modules and traits to distinguish them. To generate unique symbols that are C-compatible, the compiler mangles these to long and ugly unique names. To demangle these names back into the original, we can filter the output through [`rustfilt`](https://crates.io/crates/rustfilt). 
+
+This gives us the following stack trace (uninteresting parts removed):
+
+```plaintext
+3   ???                                 0x0000000000000000 0x0 + 0
+...
+31  libhello.dylib                      0x000000010af6e984 core::panicking::panic + 100
+32  libhello.dylib                      0x000000010af53108 core::option::Option<T>::unwrap + 56
+33  libhello.dylib                      0x000000010af4f8f3 <redis_module::alloc::RedisAlloc as core::alloc::GlobalAlloc>::alloc + 35
+34  libhello.dylib                      0x000000010af4cc8c __rg_alloc + 60
+35  libhello.dylib                      0x000000010af6e2f6 <alloc::vec::Vec<u8> as core::convert::From<&str>>::from + 38
+36  libhello.dylib                      0x000000010af4de54 <T as core::convert::Into<U>>::into + 36
+37  libhello.dylib                      0x000000010af4f57f std::ffi::c_str::CString::new + 47
+38  libhello.dylib                      0x000000010af40daa RedisModule_OnLoad + 58
+39  redis-server                        0x000000010adf97d9 moduleLoad + 118
+40  redis-server                        0x000000010adf9735 moduleLoadFromQueue + 69
+41  redis-server                        0x000000010ad94428 main + 1190
+...
+```
+
+It still took me a lot head-scratching and experimenting to figure it out, but here's what happened:
+
+The functions of the Redis modules API are accessed via C function pointers. Instead of relying on the dynamic linker to initialize these pointers, they are initialized explicitly by Redis as part of module initialization process.
+
+As the stack trace shows, during the loading of the module we call the `CString::new` function. This standard library function allocates memory for a string. This, in turn, calls our allocator which would then call `RedisModule_Alloc.unwrap()...` to actually perform the allocation. This causes a chicken-and-egg problem. The Redis module is not ready yet, meaning our function pointers have not yet been initialized, so we can't call the relevant API to perform the allocation.
+
+## The Solution
+
+I tried various approaches to solve this, but there seemed to be no clean way to avoid the allocation during module initialization. The second best thing would have been to use the standard allocator until the module is ready, and then switching to the custom one. However, Rust doesn't allow changing the allocator at runtime so we can't do that.
```

Comment:
should be "I try various approaches to solve this, but there seems to be no clean way to avoid the allocation during module initialization. The second best thing would be to use the standard allocator until the module is ready, and then switch to the custom one." to keep consistent with the present tense